rebasehelper/spec_hooks/commit_hash_updater.py

```diff
@@ -0,0 +1,121 @@
+
+    @classmethod
+    def get_name(cls):
+        return cls.NAME
+
+    @classmethod
+    def get_categories(cls):
+        return cls.CATEGORIES
+
+    @classmethod
+    def _get_commit_hash_from_github(cls, spec_file):
+        """
+        Tries to find a commit using Github API
+
+        :param spec_file: SPEC file to base the search on
+        :return: SHA of a commit, or None
+        """
+        m = re.match(r'^https?://github\.com/(?P<owner>[\w-]+)/(?P<project>[\w-]+)/.*$', spec_file.sources[0])
+        if not m:
+            return None
+        baseurl = 'https://api.github.com/repos/{owner}/{project}'.format(**m.groupdict())
+        # try to get tag name from a release matching version
+        r = requests.get('{}/releases'.format(baseurl))
+        if not r.ok:
+            if r.status_code == 403 and r.headers.get('X-RateLimit-Remaining') == '0':
+                logger.warning("Rate limit exceeded on Github API! Try again later.")
+            return None
+        data = r.json()
+        version = spec_file.get_version()
+        tag_name = None
+        for release in data:
+            if version in release.get('name'):
+                tag_name = release.get('tag_name')
+                break
+        r = requests.get('{}/tags'.format(baseurl))
+        if not r.ok:
+            if r.status_code == 403 and r.headers.get('X-RateLimit-Remaining') == '0':
+                logger.warning("Rate limit exceeded on Github API! Try again later.")
+            return None
+        data = r.json()
+        for tag in data:
+            name = tag.get('name')
+            if tag_name:
+                if name != tag_name:
+                    continue
+            else:
+                # no specific tag name, try common tag names
+                if name not in [version, 'v{}'.format(version)]:
+                    continue
+            commit = tag.get('commit')
+            if commit:
+                return commit.get('sha')
+        return None
+
+    @classmethod
+    def _get_commit_hash(cls, spec_file):
+        if 'github.com' in spec_file.sources[0]:
+            return cls._get_commit_hash_from_github(spec_file)
+        return None
+
+    @classmethod
+    def run(cls, spec_file, rebase_spec_file, **kwargs):
+        if rebase_spec_file.sources[0] != spec_file.sources[0]:
+            # nothing to do
+            return
+        # try to determine commit hash matching the new version
+        new_commit = cls._get_commit_hash(rebase_spec_file)
+        if not new_commit:
+            return
+        source = rebase_spec_file.sources[0]
+        # try to determine commit hash matching the old version
+        old_commit = cls._get_commit_hash(spec_file)
+        if old_commit:
+            # replace old commit hash with the new one
+            source = source.replace(old_commit, new_commit)
+        else:
+            # try to find anything resembling SHA1 hash
+            hashes = re.findall(r'[0-9a-f]{40}', source)
```

Comment:
Commit is searched based on release or tag name matching version. Old version of the spec file could be based on some untagged commit. In case commit is found for the new version, but not for the old, there is still a chance to succeed, if we know what to replace with the new commit hash. And if there is a single SHA1 hash somewhere in the source string, it is most probably it.

Reply:
Yes of course, that makes sense. Thanks for the insight.